### ISSUE TITLE
Failing test case with transform dialect due to potentially mis-tracked operation. 

### ISSUE THREAD

**Issue by MaheshRavishankar** [llvm/llvm-project#72931]
After using https://github.com/llvm/llvm-project/pull/72920 locally, I am having the following error with the following example :

```
builtin.module attributes { transform.with_named_sequence } {
  func.func @matmul_static_dispatch_0_matmul_128x512x256_i32(%3 : tensor<128x256xi32>, %4 : tensor<256x512xi32>) -> tensor<128x512xi32> {
    %c0 = arith.constant 0 : index
    %c0_i32 = arith.constant 0 : i32
    %5 = tensor.empty() : tensor<128x512xi32>
    %6 = linalg.fill ins(%c0_i32 : i32) outs(%5 : tensor<128x512xi32>) -> tensor<128x512xi32>
    %7 = linalg.matmul ins(%3, %4 : tensor<128x256xi32>, tensor<256x512xi32>) outs(%6 : tensor<128x512xi32>) -> tensor<128x512xi32>
    return %7 : tensor<128x512xi32>
  }

  transform.named_sequence @cleanup(%variant_op: !transform.any_op {transform.readonly}) {
    %func = transform.structured.match ops{["func.func"]} in %variant_op : (!transform.any_op) -> !transform.any_op
      transform.apply_patterns to %func {
      transform.apply_patterns.linalg.tiling_canonicalization
      transform.apply_patterns.scf.for_loop_canonicalization
      transform.apply_patterns.canonicalization
    } : !transform.any_op
    transform.apply_cse to %func : !transform.any_op
    transform.yield
  }

  transform.named_sequence @__transform_main(%variant_op: !transform.any_op {transform.read_only}) {
    %ops = transform.structured.match ops{["linalg.fill", "linalg.matmul"]} in %variant_op : (!transform.any_op) -> !transform.any_op
    %fill, %matmul = transform.split_handle %ops : (!transform.any_op) -> (!transform.any_op, !transform.any_op)

    // First level tile to forall with tile_sizes [8, 8].
    %tiled_matmul, %forall =
      transform.structured.tile_using_forall %matmul tile_sizes [8, 8]
        ( mapping = [#gpu.block<y>, #gpu.block<x>] ) : (!transform.any_op) -> (!transform.any_op, !transform.any_op)

    // Fuse fill operation into the loop
    %fused_fill, %fused_for_all = transform.structured.fuse_into_containing_op %fill into %forall : (!transform.any_op, !transform.any_op) -> (!transform.any_op, !transform.any_op)

    // Pad operation.
    %padded, %pad, %__ = transform.structured.pad %tiled_matmul {
      padding_values=[0 : i32, 0 : i32, 0 : i32],
      padding_dimensions=[0, 1, 2],
      pack_paddings=[1, 1, 1],
      copy_back_op="none"
    } : (!transform.any_op) -> (!transform.any_op, !transform.any_op, !transform.any_op)
    %pad_dps = transform.structured.rewrite_in_destination_passing_style %pad : (!transform.any_op) -> !transform.any_op

    // Promote the operands to shared memory.
    %padded_lhs = transform.get_producer_of_operand %padded[0] : (!transform.any_op) -> (!transform.any_op)
    %padded_lhs_buffer, %padded_lhs_new = transform.structured.bufferize_to_allocation %padded_lhs
        {memory_space = 1, bufferize_destination_only} : !transform.any_op

    %padded_rhs = transform.get_producer_of_operand %padded[1] : (!transform.any_op) -> (!transform.any_op)
    %padded_rhs_buffer, %padded_rhs_new = transform.structured.bufferize_to_allocation %padded_rhs
        {memory_space = 1, bufferize_destination_only} : !transform.any_op

    // Run canonicalizations.
    transform.include @cleanup failures(propagate) (%variant_op) : (!transform.any_op) -> ()

    // Find the matmul and fill again
    %tiled_ops = transform.structured.match ops{["linalg.fill", "linalg.matmul"]} in %fused_for_all : (!transform.any_op) -> !transform.any_op
    %tiled_fill_op, %tiled_padded_matmul = transform.split_handle %tiled_ops : (!transform.any_op) -> (!transform.any_op, !transform.any_op)

    // Second level tile to forall with tile_sizes [4, 4].
    %tiled_matmul_1, %forall_1 =
      transform.structured.tile_using_forall %tiled_padded_matmul tile_sizes [4, 4]
        ( mapping = [#gpu.thread<y>, #gpu.thread<x>] ) : (!transform.any_op) -> (!transform.any_op, !transform.any_op)
    %fused_fill_2, %fused_for_all_2 = transform.structured.fuse_into_containing_op %tiled_fill_op into %forall_1 : (!transform.any_op, !transform.any_op) -> (!transform.any_op, !transform.any_op)

    // Pad operation.
    %padded_1, %pad_1, %_ = transform.structured.pad %tiled_matmul_1 {
      padding_values=[0 : i32, 0 : i32, 0 : i32],
      padding_dimensions=[0, 1, 2],
      pack_paddings=[0, 0, 1],
      copy_back_op="linalg.copy"
    } : (!transform.any_op) -> (!transform.any_op, !transform.any_op, !transform.any_op)
    %pad_1_dps = transform.structured.rewrite_in_destination_passing_style %pad_1 : (!transform.any_op) -> !transform.any_op
    transform.yield
  }
}

```

The error itself does not seem to be (at least AFAICS) related to https://github.com/llvm/llvm-project/pull/72920 . Running the following command throws the following error

```
mlir-opt --transform-interpreter repro.mlir
repro.mlir:72:18: error: expensive checks failure: operation mismatch, expected linalg.copy
    %pad_1_dps = transform.structured.rewrite_in_destination_passing_style %pad_1 : (!transform.any_op) -> !transform.any_op
                 ^
repro.mlir:7:10: note: payload op: tensor.pad
    %7 = linalg.matmul ins(%3, %4 : tensor<128x256xi32>, tensor<256x512xi32>) outs(%6 : tensor<128x512xi32>) -> tensor<128x512xi32>
```

I tried to debug this, and the error happen on the last instruction 

```
  %pad_1_dps = transform.structured.rewrite_in_destination_passing_style %pad_1 : (!transform.any_op) -> !transform.any_op
```

Stepping through the transformation, I do see the underlying transformation performing the conversion to destination passing style and return  a `linalg.copy` but then when stepping through the rest of the transform dialect interpreter code, somewhere that is getting lost leading to this error. It might be that I am doing something wrong, so please do let me know if that the case.

**Comment by MaheshRavishankar** [comment]
cc @ftynse  @nicolasvasilache  or @matthias-springer  any ideas what is happening here?

**Comment by MaheshRavishankar** [comment]
Cc @Groverkss

**Comment by matthias-springer** [comment]
Apart from the memory side effects (#72986) I think there's something else going on here. #72986 should fix this issue for the moment, but I'm still investigating further...

**Comment by MaheshRavishankar** [comment]
> Apart from the memory side effects (#72986) I think there's something else going on here. #72986 should fix this issue for the moment, but I'm still investigating further...

Thanks @matthias-springer ! Let me know how else I can help

**Comment by yzhang93** [comment]
Hi @matthias-springer, I have a similar transform dialect to what @MaheshRavishankar posted. I had it working without this fix https://github.com/llvm/llvm-project/pull/72986. But now it has some error 
```
error: tracking listener failed to find replacement op during application of this transform op
      transform.apply_patterns to %func {
      ^
matmul_fill_spec_pack_peel.mlir:21:7: note: ran out of suitable replacement values
matmul_fill_spec_pack_peel.mlir:72:27: note: replacement is required because alive handle(s) exist (first use in this op as operand number 0)
    %buffer_c0, %new_c0 = transform.structured.bufferize_to_allocation %pack_producer_c0
                          ^
<stdin>:23:16: note: [0] replaced op
          %7 = linalg.matmul ins(%3, %4 : tensor<16x128xi8>, tensor<128x128xi8>) outs(%6 : tensor<16x128xi32>) -> tensor<16x128xi32>
               ^
<stdin>:23:16: note: [0] replacement value 0

```
The transform dialect script is as below
```
module attributes { transform.with_named_sequence } {
  transform.named_sequence @cleanup(%variant_op: !transform.any_op {transform.readonly}) {
    %func = transform.structured.match ops{["func.func"]} in %variant_op : (!transform.any_op) -> !transform.any_op
      transform.apply_patterns to %func {
      transform.apply_patterns.linalg.tiling_canonicalization
      transform.apply_patterns.iree.fold_fill_into_pad
      transform.apply_patterns.scf.for_loop_canonicalization
      transform.apply_patterns.canonicalization
    } : !transform.any_op
    transform.iree.apply_licm %func : !transform.any_op
    transform.iree.apply_cse %func : !transform.any_op
    transform.yield
  }

  transform.named_sequence @__transform_main(%variant_op: !transform.any_op {transform.read_only}) {
    %ops = transform.structured.match ops{["linalg.fill", "linalg.matmul"]} in %variant_op : (!transform.any_op) -> !transform.any_op
    %fill, %matmul = transform.split_handle %ops : (!transform.any_op) -> (!transform.any_op, !transform.any_op)

    // First level tile to forall with tile_sizes [16, 64].
    %tiled_matmul, %forall =
      transform.structured.tile_using_forall %matmul tile_sizes [16, 64]
        ( mapping = [#gpu.block<y>, #gpu.block<x>] ) : (!transform.any_op) -> (!transform.any_op, !transform.any_op)
    transform.iree.populate_workgroup_count_region_using_num_threads_slice %forall
      : (!transform.any_op) -> ()

    // Fuse fill operation into the forall loop.
    %fused_fill, %_ = transform.structured.fuse_into_containing_op %fill into %forall : (!transform.any_op, !transform.any_op) -> (!transform.any_op, !transform.any_op)

    // Tile reduction dimension.
    %tiled_reduction, %loop =
      transform.structured.tile_using_for %tiled_matmul [0, 0, 64]
      : (!transform.any_op) -> (!transform.any_op, !transform.op<"scf.for">)

    // Pack by applying data tiling, and the linalg.matmul becomes linalg.generic.
    %packed = transform.structured.pack %tiled_reduction packed_sizes = [16, 64, 64]
      : (!transform.any_op) -> (!transform.any_op)

    // Transpose B matrix from [K N n k] to [K N k n]
    %pack_producer_b0 = transform.get_producer_of_operand %packed[1]
      : (!transform.any_op) -> (!transform.any_op)
    %packed_b0, %pack_b0, %empty_unpack_b0 =
      transform.structured.pack_transpose %pack_producer_b0 with_compute_op(%packed)
      inner_perm = [1, 0] : (!transform.any_op, !transform.any_op)
      -> (!transform.any_op, !transform.any_op, !transform.any_op)

    // Bufferize to shared memory allocation
    %pack_producer_a0 = transform.get_producer_of_operand %packed_b0[0]
      : (!transform.any_op) -> (!transform.any_op)
    %pack_producer_c0 = transform.get_producer_of_operand %packed_b0[2]
      : (!transform.any_op) -> (!transform.any_op)
    %buffer_a0, %new_a0 = transform.structured.bufferize_to_allocation %pack_b0
      {memory_space = "shared", bufferize_destination_only, emit_dealloc} : !transform.any_op
    %buffer_b0, %new_b0 = transform.structured.bufferize_to_allocation %pack_producer_a0
      {memory_space = "shared", bufferize_destination_only, emit_dealloc} : !transform.any_op
    %buffer_c0, %new_c0 = transform.structured.bufferize_to_allocation %pack_producer_c0
      {memory_space = "shared", bufferize_destination_only, emit_dealloc} : !transform.any_op

    // Second level tile to forall with tile_sizes [1, 1].
    %tiled_matmul_1, %forall_1 =
      transform.structured.tile_using_forall %packed_b0 tile_sizes [1, 1]
        ( mapping = [#gpu.thread<y>, #gpu.thread<x>] ) : (!transform.any_op) -> (!transform.any_op, !transform.any_op)

    // Pack by applying data tiling, and the linalg.matmul becomes linalg.generic.
    %packed_2 = transform.structured.pack %tiled_matmul_1 packed_sizes = [0, 0, 0, 4, 8, 8]
      : (!transform.any_op) -> (!transform.any_op)

    // Transpose A matrix from [M K m k m0 k0] to [M K k m m0 k0]
    %pack_producer_a = transform.get_producer_of_operand %packed_2[0]
      : (!transform.any_op) -> (!transform.any_op)
    %packed_a, %pack_a, %empty_unpack_a =
      transform.structured.pack_transpose %pack_producer_a with_compute_op(%packed_2)
      outer_perm = [0, 1, 3, 2] : (!transform.any_op, !transform.any_op)
      -> (!transform.any_op, !transform.any_op, !transform.any_op)

    // Transpose B matrix from [K N k n n0 k0] to [K N n k k0 n0]
    %pack_producer_b = transform.get_producer_of_operand %packed_a[1]
      : (!transform.any_op) -> (!transform.any_op)
    %packed_b, %pack_b, %empty_unpack_b =
      transform.structured.pack_transpose %pack_producer_b with_compute_op(%packed_a)
      outer_perm = [0, 1, 3, 2] inner_perm = [1, 0] : (!transform.any_op, !transform.any_op)
      -> (!transform.any_op, !transform.any_op, !transform.any_op)

    // Transpose C matrix from [M N m n m0 n0] to [M N n m m0 n0]
    %unpack = transform.get_consumers_of_result %packed_b[0]
      : (!transform.any_op) -> (!transform.any_op)
    %packed_c, %pack_c, %unpack_c =
      transform.structured.pack_transpose %unpack with_compute_op(%packed_b)
      outer_perm = [0, 1, 3, 2] : (!transform.any_op, !transform.any_op)
      -> (!transform.any_op, !transform.any_op, !transform.any_op)

    // Bufferize to local memory allocation
    %buffer_a, %new_a = transform.structured.bufferize_to_allocation %pack_a
      {memory_space = "local", bufferize_destination_only} : !transform.any_op
    %buffer_b, %new_b = transform.structured.bufferize_to_allocation %pack_b
      {memory_space = "local", bufferize_destination_only} : !transform.any_op
    %buffer_c, %new_c = transform.structured.bufferize_to_allocation %pack_c
      {memory_space = "local", bufferize_destination_only} : !transform.any_op

    // Hoist static alloc out of the loops
    %memref_func = transform.structured.match ops{["func.func"]} in %variant_op
      : (!transform.any_op) -> !transform.any_op
    transform.iree.hoist_static_alloc %memref_func : (!transform.any_op) -> ()

    // Peel the first iteration out of the for loop
    %1 = transform.get_parent_op %packed_c {op_name = "scf.for"} : (!transform.any_op) -> !transform.op<"scf.for">
    %main_loop, %remainder = transform.loop.peel %1 {peel_front = true}
      : (!transform.op<"scf.for">) -> (!transform.any_op, !transform.any_op)
    transform.include @cleanup failures(propagate) (%variant_op) : (!transform.any_op) -> ()
    transform.yield
  }
}
```
It seems `transform.apply_patterns.canonicalization` fails after the peel transform I made in https://github.com/llvm/llvm-project/pull/74015/. But I think the real cause is the `transform.structured.bufferize_to_allocation` as shown in the error message. And if I revert https://github.com/llvm/llvm-project/pull/72986, the error is gone.

Could you please take a look and give some insight of this issue?

**Comment by matthias-springer** [comment]
I have a suspicion what's going on. The issue probably disappears if you inline the `@cleanup` function. I think the handle liveness analysis does not work with functions at the moment, so it thinks that `%pack_producer_c0` is still alive.

#72986 removes a debugging check. It should not affect the behavior. The fact that #72986 changes the behavior shows that the debug check was implemented incorrectly and removes payload ops from handles that should not have been removed. (That was one of the reasons why I removed it.) In other words, you should have seen this error all the time. The proper way to fix this is to improve the liveness check. I'm going to work on that.

Do you have way to reproduce this just in MLIR without IREE? If not how do I run reproduce this with IREE? (And I also need the input IR that the transform script operates on.)

**Comment by matthias-springer** [comment]
Can you check if this fixes the issue: #74290

**Comment by Abhishek-Varma** [comment]
Hi @matthias-springer ,

1. Here is the minimal repro we have - [repro_transform.mlir](https://gist.github.com/Abhishek-Varma/c0d708219a5f3b7f263790ffe0697394).

2. I took the IR right before the "cleanup" is being applied and made another repro - [repro_transform2.mlir](https://gist.github.com/Abhishek-Varma/41ea5bee3e136ef13c2c218e31a517a9) - and this one successfully passes - you can check the OUTPUT. 
So, your suspicion of inlining "cleanup" function is indeed correct.

3. I therefore cherry-picked your patch and re-run the minimal repro (point 1) and confirm that this solves the issue.

4. I also applied it to the IREE specific script and confirm that it works now! Thanks for this quick fix!

**Comment by yzhang93** [comment]
> Can you check if this fixes the issue: #74290

Thanks @matthias-springer for your quick fix! I applied it on my transform dialect example and confirm it works!
